### DISCUSSION TITLE
Oqtane 4.0.0 / .NET 7 Page Resource Improvements

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2826] [General]
One of the core philosophies for Oqtane (https://www.oqtane.org/blog/!/20/oqtane-philosophy) is Microsoft Alignment... which essentially means taking advantage of the native solutions offered by Blazor as opposed to creating custom solutions ie. "don't reinvent the wheel". The upgrade to .NET 7 provided the opportunity to revisit some features which had been implemented using custom solutions in Oqtane due to limitations in earlier versions of Blazor. 

The initial area of focus is related to the management of CSS and JavaScript references. These types of references generally need to be included in the head section of an HTML page. In a typical Blazor application, developers are encouraged to hard-code the references in their root document (usually _Host.cshtml if using the ASP.NET Hosted approach). However in a framework like Oqtane, where modules are developed and deployed independently from the framework and composed dynamically at run-time to create a functional application, it is not possible for developers to modify the root document. Therefore, a capability needed to be included which allowed developers to dynamically inject content into the root document. 

One very important thing to keep in mind is that Oqtane is a single page application. In a single page application, the initial request is processed by the server however all subsequent requests are processed by the client browser. What this means is that although it may appear that you are navigating from one page to another when you click navigation links in a page, there is never a full page refresh of content being delivered from the server - the content is being fully managed client-side in the browser. So basically as a developer you need to manage two distinct and separate workloads - a workload which initially runs on the server to construct the initial content for the page, and a workload which runs in the browser and modifies the content based on user interaction. These workloads often require completely different approaches to accomplish the same task.

A good example of this is managing resources in the head of the page. The initial request for a page is processed server-side and CSS and JavaScript references can easily be included in the head section of the Razor page output and sent to the browser. However once the application is running in the browser and a user invokes an action which requires an additional resource to be loaded in the head section, you are faced with a challenge. This is because Blazor traditionally did not allow you to manipulate the head of the document client-side - your main app component was in the page body. As a result you needed to use JavaScript methods to manipulate the head which required the use of JS Interop. With CSS references, this quickly became even more complicated as they require a very precise ordering on a page to support their cascading behavior. In order to allow users to have different themes and modules on each page of a site (a fundamental requirement in Oqtane), a custom solution had to be developed to support this client-side scenario (for .NET Core 3.1 in early 2020). 

This solution worked well for many years but with the arrival of .NET 7, Blazor introduced a couple of new features which promised to offer native support for this scenario. First, was the ability to have multiple root components in your default document. This was introduced in .NET 6 and allowed you to include a root component inside the head section which could be rendered dynamically. Second, was an overhaul of the pre-rendering in Blazor in .NET 7. This was critical, as in .NET 6 root components were executed in series which created major challenges if there were dependencies between them (which is extremely common as your main app component typically contains your router which is the brain of your entire application and often needs to interface with other root components). These new capabilities made it possible to replace the custom head resources solution in Oqtane with a native .NET 7 solution.

This was accomplished by introducing a new Head component and using the existing INotifyPropertyChanged capability in Oqtane (https://www.oqtane.org/blog/!/40/inotifypropertychanged) to communicate with the App component. This solution was simple and allowed workloads which had previously been implemented on both the server and client to be consolidated - improving performance and simplifying the code. Some new methods were also added to ModuleBase to make it easy for developers to programmatically modify the title of the page and include head content (ie. meta, link, script tags) on both the client and server - capabilities which were not possible previously. 

Now some might ask, if the core philosophy is Microsoft Alignment, why not use the HeadOutlet approach which is included in a standard .NET 7 Blazor template project? One reason is because when it was introduced in .NET 6 it did not function as intended which required you to restructure your base Blazor project in order to use it (non-trivial in Oqtane). In .NET 7 this issue was finally resolved (as I described above). However the HeadOutlet component is also tightly coupled with the default Blazor "routable component" and router approach which does not align well with Oqtane's modular component approach.

Last year an enhancement was added to Oqtane to allow an Administrator to specify Meta tags for a page. Although this feature worked as originally intended, it did not support a variety of other related use cases. For example, it did not allow an Administrator to manage Meta tags at a site level. Nor did it support other head elements such as links or scripts. In order to resolve these limitations, a new property has been introduced at the Site level for HeadContent and the former Meta property at the Page level was renamed to HeadContent. These enhancements rely on the new root component concept described above. However during testing, I realized that script elements defined at the page level are problematic. They work fine on the initial page request but if they are injected as a subsequent client-side interaction they do not behave as expected.

During pre-rendering on the server, the root component will render script elements, and when they are received by the browser they will be executed as expected. However once the Blazor application becomes interactive, the Head component executes client-side. In this scenario if script elements are being injected into the head, they will be added to the DOM but they are not executed. This is a well known limitation of Blazor, as it can only manage JavaScript through JS Interop. 

The main question is if it is critical for Administrators to be able to include scripts at the page level - or if it is sufficient to only provide this capability at the site level? (site level script injection works fine because it occurs during pre-rendering on the server)

In order to support scripts at the page level, the HeadContent would need to be parsed and JS Interop would need to be used to dynamically manage the scripts in the page. It turns out that this is quite a bit more complicated than it sounds due to the fact that that there is no reliable way on both Blazor and Blazor WebAssembly to determine if a component is in the pre-rendering phase of the life cycle. So essentially you need to exclude the scripts from being rendered by the Head component and use traditional server page rendering techniques, and also implement client-side rendering using JS Interop. Essentially it is the same legacy custom approach which I was hoping to avoid based on the new capabilities in .NET 7. Note that other head elements such as meta tags or link tags do not have the same technical challenges as scripts and can be supported at the page level without difficulty. 

*** So to clarify, I am looking for some feedback from the community on whether or not it would be acceptable from a functionality perspective if Oqtane restricted **script elements** from being included in the  in the **Head Content** property at the **Page level**? 

(Note that the current code in the repo - https://github.com/oqtane/oqtane.framework/tree/0c749a126c02a4defb13e63706e796b8868f2dc9 - does support scripts at the page level - however the code and flow is very complicated and the parsing of script elements is fragile)

There are still a few other items outstanding to fully complete the root document enhancements...

One item is how to provide support for items which need to be added to the body of the page. Script references often need to be added to the end of the body section rather than the head section. If we can agree on the proposal above that scripts should only be allowed at the site level it reduces the scope significantly. The obvious approach would be to add another property at the site level for Body Content which can be specified by an Administrator in the UI using free-form text (exactly the same as Head Content). 

The other item is how to replace IHostResources with a capability which offers the same benefits, but without the current drawbacks. Note that the addition of the Head Content and Body Content properties at the site level mentioned previously would already satisfy some of the use cases which IHostResources were used for. However they use a configuration approach rather than a programmatic approach. The thought here is that instead of having an interface which is not tied to any of the other framework semantics, it would be better if the IModule and ITheme interfaces included properties for Resources. Then it would be possible for developers to identify the resources which should be used for their entire module or theme, and not have to repeat them in each component (although this would still be possible for backward compatibility as well as the cases where more granularity is needed). By associating resources with a module or theme, the framework will be able to determine which resources are related to a specific site. And it would be possible to indicate that resources are local or global in scope - global resources would be included for the entire site, whereas local resources would be injected for the specific page where a module or theme is utilized. 

Thoughts/feedback are appreciated...

**Comment by thabaum** [comment]
Since Oqtane is a SPA, I think that site level Body Content should be fine for most use cases.  Would there be possibility of an issue if you had a page that did not need a script loaded at site level that may conflict with one needed specially for a only site page(s) a module may need Body Content property to load?

I am just thinking of if you had a number of different module vendors loading scripts in Body Content would there be a chance for dependency conflicts, naming conflicts and order of script loading.  If so how can we help to mitigate these types of potential script conflicts.

**Reply by sbwalker** [reply]
@thabaum Oqtane will ensure duplicate script references are not added to the head or body

**Comment by sbwalker** [comment]
#2828 adds the ability to define Resources in the IModule and ITheme interfaces

**Reply by vnetonline** [reply]
Oqtane **should allow** script elements from being included in the in the Head Content property at the Page level?

Because you may need to have specific scripts which are just needed for that specific landing page for example a dashboard which requires script to move components around the page dynamically

**Comment by vnetonline** [comment]
Well done @sbwalker elegant solution very much looking forward to use it

**Comment by sbwalker** [comment]
I believe there is still a technical problem when it comes to scripts - and it's a problem which exists in all of the prior Oqtane releases as well. 

When using scripts in a traditional server-based application such as MVC or WebForms, the script payload in the page is always executed by the browser. This is because the browser gets a full page refresh of content on each request. 

Currently in Oqtane, scripts are executed on the first page request. And as you navigate from page to page in the browser it will dynamically inject scripts client-side as needed. However the way it injects scripts is by determining if they already exist in the DOM. If they have already been injected previously then they are not added again. However this also means they are not re-executed. So this is not the same behavior as you would get in a traditional server-side application. 

I believe the expected behavior is that scripts should mimic the traditional server-side application behavior and re-execute on each page. In order to do this, it would actually need to replace the script in the DOM if it already exists. However I have not had much success in removing elements using JavaScript - it throws Blazor errors. I will continue to investigate.

**Reply by hishamco** [reply]
> I believe the expected behavior is that scripts should mimic the traditional server-side application behavior and re-execute on each page. In order to do this, it would actually need to replace the script in the DOM if it already exists. However I have not had much success in removing elements using JavaScript - it throws Blazor errors. I will continue to investigate.

Maybe I saw this late, could you please explain what is the exact issue, I might help with my humble front-end skills

**Reply by sbwalker** [reply]
@hishamco let's consider a very common use case - adding a Google Analytics script to your site. In a traditional web app the script is included in the output of each page request sent to the browser, and the script is able to obtain information about the current page url, etc...

Now let's examine the same scenario for a Blazor app (and let's assume that it is configured for pre-rendering on the server). The initial request to the server will include the script in the output and the browser will receive it and execute it. The user clicks on Page2 in the navigation. The action is handled client-side. Note the script already exists in the page from the original request - but it is now inactive.. In order to execute the script the Blazor app needs to use JS Interop to add the script to the page again.  It is possible to simply inject the script again... and it will be executed. However now you will have 2 versions of the same script in the page. Now click on Page3... and you now have 3 versions of the script in the page... etc... To prevent the page from filing up with inactive scripts, it makes sense that you should remove the script from the page using JS Interop before adding it again. However Blazor does not like when you use the standard GetElementById and Remove methods in JavaScript - it will throw errors in the console and freeze your application. My guess is that it is due to the way that Blazor disposes of components and removing elements creates a mismatch between the page and the shadow DOM. I have not been able to track this down yet.

**Reply by sbwalker** [reply]
@hishamco I found a solution today... sometimes it helps to describe the problem in detail to get the clarity to solve it :)

**Reply by hishamco** [reply]
How you solve it?

**Reply by vnetonline** [reply]
I am also interested in how did you solve it ?

**Reply by sbwalker** [reply]
I will submit a new PR this week and provide an explanation

**Reply by iJungleboy** [reply]
@sbwalker Just FYI: The way we're doing Analytics today is that the theme loads the JS, and on following pages it interops to report another page view. Easy to do when you know your js/analytics, but not as simple as "just-add-script"

**Comment by iJungleboy** [comment]
TBH: I don't fully understand the question.

ATM you can add scripts on each module and theme. And adding them at site or theme-level makes sense too. In the Page-Scenario an optional additional variant, or would no-page also mean no-module?

**Comment by sbwalker** [comment]
@iJungleboy its hard to explain so I have been using a table to track all of the Script use cases. The table contains all of the possible ways that script can be defined in Oqtane 4.0.0. The items highlighted in yellow are new capabilities which did not exist previously. The item in orange is deprecated in favor of the new methods:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/ad7f96e3-7729-423e-872f-4114c68d4dc1)

The table identifies if the script is handled on the Server or Client... and if its on the Client it identifies which component is responsible and which Interop method.

**Reply by sbwalker** [reply]
@iJungleboy the remaining challenge is that .NET MAUI does not support a remote host (ie. no _Host.cshtml) - it only supports a locally installed index.html. So in all of the cases above where I indicated that scripts are being handled on the Server, there needs to be logic to detect if the client .NET MAUI and if so, inject the scripts on the Client.

**Reply by iJungleboy** [reply]
@sbwalker is `internal` the same as `inline` or on-site/off-site?

**Reply by sbwalker** [reply]
internal = script tag with no src attribute
external = script tag with src attribute

**Comment by iJungleboy** [comment]
@sbwalker Just as a possible thought and inspiration: The scripts have 2 challenges (don't want to make things harder...)

1. Including it and parsing it
2. Calling it - possibly with parameters that were server generated; ideally after verifying that all dependencies are loaded

Especially point 2 was something I worked on a lot for 2sxc. Here's just the gist:

* in DNN everything used to just run if it was inline, so parsed etc. - this doesn't work on Oqtane of course
* But then I also wanted to solve some CSP issues with signed scripts etc. - which doesn't work if your script has dynamic parameters such as a module-id (eg. a gallery which should be initialized, but only the images on module 7402 should be together...)
* Since all 2sxc code was meant to be hybrid (same App eg. Blog should run on Dnn and Oqtane without modifications) I had to find a convention that always works
* I created a system called `turnOn` https://github.com/2sic/turnOn

It basically takes a configuration such as this:

```js
const config = {
  run: "window.myApp.Gallery.init()",
  require: ['window.fancybox'],
  data: {
    id: 7402
  }
}
```

There's a bit of magic because you can place this in a special HTML tag which is auto-picked up; and it will also check that `window.myApp` even exists before it triggers the `.init()` and I don't want to bore you with the details. 
Instead I just wanted to possibly give you some ideas for this scenario, because it helped us create a very robust mechanism to ensure that no matter when scripts were loaded, timing issues - before or after the HTML that is needed - everything just works.

If you're interested we can do a telco or something, just reach out.

**Reply by sbwalker** [reply]
All of the use cases I described are already implemented and working - the only item remaining is the .NET MAUI integration. One item which is not explicitly outlined in the use cases above is that it is now possible to include dynamic script in modules and themes. For example, a module could include the following in the OnInitialized() or OnParametersSet() method using standard C# string interpolation:

```
AddScript(new Resource { Content = $"console.log(\"Hello {ModuleState.Title}!!\");" });
```
And this would inject the script at runtime.  Note that this is simply an extension of the legacy ability to specify Resources statically for a component during its initialization.

In regards to dependency management, the LoadJS library (https://github.com/muicss/loadjs) which Oqtane has used since version 1 handles this scenario for registering external JavaScript libraries. You simply need to specify the Bundle parameter for each resource that are dependent on one another (there is an example of this in the rich text editor).
